neuraxle/metaopt/random.py

```diff
@@ -114,6 +117,278 @@ def _split(self, data_inputs):
+
+        :param data_inputs: data to perform walk forward cross validation into.
+        :param expected_outputs: the expected target*label that will be used during walk forward cross validation.
+        :return: train_data_inputs, train_expected_outputs
+        """
+        splitted_data_inputs = self._train_split(data_inputs)
+        if expected_outputs is not None:
+            splitted_expected_outputs = self._train_split(expected_outputs)
+            return splitted_data_inputs, splitted_expected_outputs
+
+    def validation_split(self, data_inputs, expected_outputs=None) -> List:
+        """
+        Split the data into validation inputs, validation expected outputs.
+
+        Notes: The data supported by this cross validation is nd.array of shape = (batch, timeSeries, ...) .
+        The array can have an arbitrary number of dimension, but the time series axis is currently limited to `axis=1`.
+        Recommended format for the data is an numpy array of shape [batch_size, total_time_steps, n_features].
+
+        :param data_inputs: data to perform walk forward cross validation into.
+        :param expected_outputs: the expected target*label that will be used during walk forward cross validation.
+        :return: validation_data_inputs, validation_expected_outputs
+        """
+        splitted_data_inputs = self._validation_split(data_inputs)
+        if expected_outputs is not None:
+            splitted_expected_outputs = self._validation_split(expected_outputs)
+            return splitted_data_inputs, splitted_expected_outputs
+        return splitted_data_inputs
+
+    def _train_split(self, data_inputs):
+        splitted_data_inputs = []
+        number_step = self._get_number_fold(data_inputs)
+
+        for i in range(number_step):
+            # first slice index is always 0 for anchored walk forward cross validation.
+            a = 0
+            b = int(self.minimum_training_size + i * self.validation_window_size)
+
+            if b > data_inputs.shape[1]:
+                b = data_inputs.shape[1]
+
+            # TODO: The slicer could a inverse_transform of the joiner. A len method should also be defined.
+            slice = data_inputs[:, a:b]
+            splitted_data_inputs.append(slice)
+        return splitted_data_inputs
+
+    def _validation_split(self, data_inputs):
+        splitted_data_inputs = []
+        number_step = self._get_number_fold(data_inputs)
+        for i in range(number_step):
+            a = int(self._validation_initial_start + i * self.validation_window_size)
+            b = int(a + self.validation_window_size)
+
+            if b > data_inputs.shape[1]:
+                b = data_inputs.shape[1]
+
+            # TODO: The slicer could a inverse_transform of the joiner. A len method should also be defined.
+            slice = data_inputs[:, a:b]
+            splitted_data_inputs.append(slice)
+        return splitted_data_inputs
+
+    def _get_number_fold(self, data_inputs):
```

Comment:
This seems like duplicated code. The ISP and LSP wouldn't be broken if both classes inherited from a common other class. This would avoid duplication of perhaps a few comments, and of a few functions.

Reply:
I am not sure what do I do with the minimum_training_size of the anchored version and training_size attribute of the classic version? Especially with your other comments that you want training_size to be mandatory and not optional (comments about training comes before validation).

Here is the solution, I think of :
- Create a base WalkForwardCrossValidation class that anchored and classic version inherits from.
- Rename it both minimum_training_size (might lose clarity for the classic version).
- Rename it both training_size (might lose clarity for the anchored version).
- Overwrite the anchored init constructor calling for a minimum_training_size but set the training_size attribute of the super in the background.
- Overwrite the anchored init constructor without calling the super().__init__() from the classic version. It will however loose the MRO.
- Do not pass all parameters to the super().__init__(). Which seems to be an antipattern:  https://docs.quantifiedcode.com/python-anti-patterns/correctness/missing_argument_to_super.html

If you have any other idea, I am open to it. Which solution do you prefer I implement?

Reply:
Good for creating a WalkForwardCrossValidation class that the two other inherit from! 

This could be done in another PR as well, as the API of the two derivated classes doesn't change much from this change in abstraction.